src/ray/gcs/gcs_server/gcs_placement_group_manager.h

```diff
@@ -78,5 +78,88 @@ class GcsPlacementGroup {
+  /// `registered_placement_groups_` and its state is `ALIVE`. The callback will not be
+  /// called in this case.
+  Status RegisterPlacementGroup(const rpc::CreatePlacementGroupRequest &request,
+                                RegisterPlacementGroupCallback callback);
+
+  /// Get the placement_group ID for the named placement_group. Returns nil if the
+  /// placement_group was not found.
+  /// \param name The name of the  placement_group to look up.
+  /// \returns PlacementGroupID The ID of the placement_group. Nil if the
+  /// placement_group was not found.
+  PlacementGroupID GetPlacementGroupIDByName(const std::string &name);
+
+  /// Schedule placement_groups in the `pending_placement_groups_` queue.
+  void SchedulePendingPlacementGroups();
+
+  /// Handle placement_group creation task failure. This should be called when scheduling
+  /// an placement_group creation task is infeasible.
+  ///
+  /// \param placement_group The placement_group whose creation task is infeasible.
+  void OnPlacementGroupCreationFailed(std::shared_ptr<GcsPlacementGroup> placement_group);
+
+  /// Handle placement_group creation task success. This should be called when the
+  /// placement_group creation task has been scheduled successfully.
+  ///
+  /// \param placement_group The placement_group that has been created.
+  void OnPlacementGroupCreationSuccess(
+      std::shared_ptr<GcsPlacementGroup> placement_group);
+
+  void HandleCreatePlacementGroup(const rpc::CreatePlacementGroupRequest &request,
+                                  rpc::CreatePlacementGroupReply *reply,
+                                  rpc::SendReplyCallback send_reply_callback) override;
+
+  void ScheduleTick();
+
+ private:
+  /// Callbacks of placement_group registration requests that are not yet flushed.
+  /// This map is used to filter duplicated messages from a Driver/Worker caused by some
+  /// network problems.
+  absl::flat_hash_map<PlacementGroupID, std::vector<RegisterPlacementGroupCallback>>
+      placement_group_to_register_callbacks_;
+  /// All registered placement_groups (pending placement_groups are also included).
+  absl::flat_hash_map<PlacementGroupID, std::shared_ptr<GcsPlacementGroup>>
+      registered_placement_groups_;
+  /// Maps detached placement_group names to their placement_group ID for lookups by name.
+  absl::flat_hash_map<std::string, PlacementGroupID> named_placement_groups_;
+  /// The pending placement_groups which will not be scheduled until there's a resource
+  /// change.
+  std::deque<std::shared_ptr<GcsPlacementGroup>> pending_placement_groups_;
+  /// The scheduler to schedule all registered placement_groups.
+  std::shared_ptr<gcs::GcsPlacementGroupSchedulerInterface>
+      gcs_placement_group_scheduler_;
+  /// Used to update placement group information upon creation, deletion, etc.
+  std::shared_ptr<gcs::GcsTableStorage> gcs_table_storage_;
+  /// If a placement group is creating
+  bool is_creating_ = false;
```

Comment:
Isn't this the same as !pending_placement_groups_.empty()?

Reply:
No, It will be true when the pending_placement_groups_.empty() == false but there is a placement group is creating now.